yield_plugin/src/yield_plugin.cpp

```diff
@@ -37,72 +37,344 @@ using oss = std::ostringstream;
+          received_cooperative_request_ = true;
+          double req_plan_time = req_expiration_sec - current_time;
+          set_incoming_request_info(req_traj_plan, req_traj_speed, req_plan_time);
+          lc_status_msg.status = cav_msgs::LaneChangeStatus::REQUEST_ACCEPTED;
+          lc_status_msg.description = "Cooperative Lane Change Request Accepted by Yield Plugin";
+        
+        }
+        else
+        {
+          ROS_DEBUG_STREAM("Igonore expired Mobility Request.");
+          outgoing_response.is_accepted = false;
+          lc_status_msg.status = cav_msgs::LaneChangeStatus::REQUEST_REJECTED;
+          lc_status_msg.description = "Cooperative Lane Change Request Rejected by Yield Plugin";
+        }
+        mobility_response_publisher_(outgoing_response);
+        lc_status_msg.status = cav_msgs::LaneChangeStatus::RESPONSE_SENT;
+        lc_status_msg.description = "Cooperative Lane Change Response Sent";
+      }
+    }
+    if (lanechange_status_pub_.isLatched())
+    {
+      lanechange_status_pub_.publish(lc_status_msg);
+    }
+    else
+    {
+      ROS_WARN("Lane Change Status Topic is not latched.")
+    }
+    
+  }
+
+  void YieldPlugin::set_incoming_request_info(std::vector <lanelet::BasicPoint2d> req_trajectory, double req_speed, double req_planning_time)
+  {
+    req_trajectory_points_ = req_trajectory;
+    req_target_speed_ = req_speed;
+    req_target_plan_time_ = req_planning_time;
+  }
+
+
+  void YieldPlugin::bsm_cb(const cav_msgs::BSMConstPtr& msg)
+  {
+    cav_msgs::BSMCoreData bsm_core_ = msg->core_data;
+    host_bsm_id_ = bsmIDtoString(bsm_core_);
+  }
+
   bool YieldPlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest& req,
                                               cav_srvs::PlanTrajectoryResponse& resp)
   {
     if (req.initial_trajectory_plan.trajectory_points.size() < 2){
       throw std::invalid_argument("Empty Trajectory received");
     }
     cav_msgs::TrajectoryPlan original_trajectory = req.initial_trajectory_plan;
-    cav_msgs::TrajectoryPlan trajectory;
-    trajectory.header.frame_id = "map";
-    trajectory.header.stamp = ros::Time::now();
-    trajectory.trajectory_id = original_trajectory.trajectory_id;
+    cav_msgs::TrajectoryPlan yield_trajectory;
 
-    trajectory = update_traj_for_object(original_trajectory, req.vehicle_state.longitudinal_vel); // Compute the trajectory
-    trajectory.initial_longitudinal_velocity = original_trajectory.initial_longitudinal_velocity;//copy the original trajectory's desired speed for now. 
+    // seperating cooperative yield with regular object detection for better performance.
+    if (config_.enable_cooperative_behavior && received_cooperative_request_)
+    {
+      tf2_listener_.reset(new tf2_ros::TransformListener(tf2_buffer_));
+      yield_trajectory = update_traj_for_cooperative_behavior(original_trajectory, req.vehicle_state.longitudinal_vel);
+      // reset the flag
+      received_cooperative_request_ = false;
```

Comment:
If we reset this flag every timestep, are we in danger of having some strange behavior? I'm concerned if they're a little late next timestep with sending us the message, will that cause us to not yield or to kind of toggle in and out of yield behavior on every other timestep?

Reply:
if the request is not received, the yield plugin performs its regular object avoidance logic.

Reply:
That makes sense, but in the case where we're slowing down to avoid a future collision or to make space for the merging vehicle (so object avoidance would not yet be triggered) running alternating trajectories of regular behavior and yield could produce some strange results. 

Another possible consideration here is network latency and bandwidth, a full negotiation cycle (mobility request -> mobility ack, TX and RX for both) can take several (potentially 3-5 timesteps) to complete even in nominal conditions. If we're relying on completing a handshake on lane change plan every timestep we may be setting ourselves up for issues, especially as we scale out the number of participants in the network.